zerver/lib/test_classes.py

```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
 
     def setUpBeforeMigration(self, apps: StateApps) -> None:
         pass  # nocoverage
+
+class SlapdObject(slapdtest.SlapdObject):
+    # This class will create some temporary directories and files
+    # under TMPDIR. We override it to /tmp, since otherwise
+    # it uses the current working directory.
+    TMPDIR = "/tmp"
+
+    def setup_rundir(self) -> None:
+        """
+        Override to add our own schemas
+        """
+        super().setup_rundir()
+
+        ldap_fixture_dir = os.path.join(
+            os.path.dirname(__file__),
+            "../tests/fixtures/ldap/"
+        )
+        # Create a link to the schema in slapdtest's internal directory:
+        self._ln_schema_files(['user_extension.schema'], ldap_fixture_dir)
+        # Add to the openldap_schema_files list, so that it gets imported:
+        self.openldap_schema_files.append('user_extension.schema')
+
+class SlapdTestCase(ZulipTestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        """
+        This setup code is quite expensive, so it should be run once
+        for the entire class, rather than having it in normal setUp()
+        and running it for every single test here.
+        """
+        super().setUpClass()
+
+        cls.server = SlapdObject()
+        cls.server.suffix = "dc=zulip,dc=com"
+        cls.server.openldap_schema_files = [
+            "core.schema",
+            "cosine.schema",
+            "inetorgperson.schema",
+        ]
+
+        cls.server.start()
+
+        ldif = cls.fixture_data('directory.ldif', type='ldap')
+        cls.server.ldapadd(ldif)
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        super().tearDownClass()
+        cls.server.stop()
+
+    def settings(self, **kwargs: Any) -> Any:
+        kwargs["AUTH_LDAP_SERVER_URI"] = self.server.ldap_uri
+        if "AUTH_LDAP_USER_SEARCH" not in kwargs:
+            kwargs["AUTH_LDAP_USER_SEARCH"] = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                         ldap.SCOPE_SUBTREE, "(uid=%(user)s)")
+        if not kwargs.get("LDAP_APPEND_DOMAIN"):
+            # If LDAP_APPEND_DOMAIN is not used, email search should be configured.
+            # Use defaults here, unless the test specifies its own settings for these.
```

Comment:
Let's explain this more clearly: "If LDAP_APPEND_DOMAIN is specified, then AUTH_LDAP_USER_SEARCH can be used to look up users by email address.  Otherwise, we need to declare how to look up users by email address."

Reply:
Added